torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -31,8 +31,203 @@ class ChunkDataReader {
+    example_sampler_.reset(data_size);
+
+    if (!batch_queue_.empty()) {
+      // if the queue has exiting data, and the last batch doesn't have enough
+      // examples to fill a batch_size batch, add more example to this batch first.
+      auto& batch_data = batch_queue_.back();
+      size_t current_count = batch_data.batch_data.size();
+      if (current_count < batch_size_) {
+        auto example_count =
+            std::min(remaining_size, batch_size_ - current_count);
+        auto batch_example_indices = example_sampler_.next(example_count);
+        AT_ASSERT(
+            batch_example_indices &&
+            batch_example_indices.value().size() == example_count)
+        BatchRequestType indices = batch_example_indices.value();
+        for (size_t i : indices) {
+          batch_data.batch_data.emplace_back(std::move(data[i]));
+        }
+        remaining_size -= example_count;
+      }
+    }
+
+    // If we still have data remaining after filling the last pushed batch, add
+    // them to the queue too.
+    while (remaining_size > 0) {
+      UnwrappedBatchType current_batch;
+
+      // Allocate the batch memory ahead of time.
+      current_batch.reserve(batch_size_);
+
+      auto example_count = std::min(remaining_size, batch_size_);
+      auto batch_example_indices = example_sampler_.next(example_count);
+      AT_ASSERT(
+          batch_example_indices &&
+          batch_example_indices.value().size() == example_count)
+      BatchRequestType indices = batch_example_indices.value();
+      for (size_t i : indices) {
+        current_batch.emplace_back(std::move(data[i]));
+      }
+      remaining_size -= example_count;
+      UnwrappedBatchData<UnwrappedBatchType> batch_data(
+          std::move(current_batch));
+      batch_queue_.push(std::move(batch_data));
+    }
+    this->total_example_count_in_queue_ += data_size;
+
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    this->remaining_chunk_count_--;
+
+    lock.unlock();
+    cvr_.notify_all();
+  }
+
+  /// Push exceptions throwed during preloading into batch queue. Called from
+  /// the ChunkDataSet worker threads.
+  void add_chunk_data(size_t index, std::exception_ptr e_ptr) {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cvw_.wait(lock, [this] {
+      // stop loading if we have preloaded enough data.
+      return this->total_example_count_in_queue_ < queue_depth_;
+    });
+    UnwrappedBatchData<UnwrappedBatchType> batch_data(e_ptr);
+    batch_queue_.push(std::move(batch_data));
+
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    remaining_chunk_count_--;
+    lock.unlock();
+    cvr_.notify_all(); // notify all readers.
+  }
+
+  /// count of remaining chunk to be loaded. It is initialized with the total
+  /// chunk count and it decreases when a chunk data is retrieved. When this reaches
+  /// to 0, no more chunk needs to be loaded.
+  size_t remaining_chunk_count_ = 0;
+
+  size_t batch_size_ = 0;
```

Comment:
A comment like this.
/// The batch size is needed to create batches from the chunk data. Similar to regular dataloader where the batches are created with prefetches, BatchDataBuffer perfom the batch creation using the provided batch size.